db/manual_compaction_test.cc

```diff
@@ -60,6 +47,58 @@ class DestroyAllCompactionFilter : public CompactionFilter {
   const char* Name() const override { return "DestroyAllCompactionFilter"; }
 };
 
+class LogCompactionFilter : public CompactionFilter {
+ public:
+  const char* Name() const override { return "LogCompactionFilter"; }
+
+  bool Filter(int level, const Slice& key, const Slice& /*existing_value*/,
+              std::string* /*new_value*/,
+              bool* /*value_changed*/) const override {
+    key_level_[key.ToString()] = level;
+    return false;
+  }
+
+  void Reset() { key_level_.clear(); }
+
+  int NumKeys() const { return key_level_.size(); }
+
+  bool KeyExists(const Slice& key) {
+    return key_level_.find(key.ToString()) != key_level_.end();
+  }
+
+  int KeyLevel(const Slice& key) {
+    auto it = key_level_.find(key.ToString());
+    if (it == key_level_.end()) {
+      return -1;
+    }
+    return it->second;
+  }
+
+ private:
+  mutable std::map<std::string, int> key_level_;
+};
+
+class ManualCompactionTest : public testing::Test {
+ public:
+  ManualCompactionTest() {
+    // Get rid of any state from an old run.
+    dbname_ = test::PerThreadDBPath("rocksdb_manual_compaction_test");
+    DestroyDB(dbname_, Options());
+  }
+
+  void AssertCompacted(
```

Comment:
although this makes the test more dense, I'm going to remove this because on test failure, this cannot show clearly where the test fails, the failure will always look like ```db/manual_compaction_test.cc:92: Failure
Expected equality of these values:
  key_levels.size()
    Which is: 1
  filter->NumKeys()
    Which is: 2``` instead of pointing to the exact location.